### PR TITLE
[orc8r][sms] Add SQL storage implementation for cloud smsd

### PR DESCRIPTION
**Pull request by xjtian**
Signed-off-by: Jacky Tian <xjtian@fb.com>

## Summary

- SQL storage implementation for SMS cloud service that will track SMS's to be delivered to AGW's
- Due to unfortunate realities in the implementation of SMS, this is way more complicated than it should be. The main bulk of the complexity lies around the "reference numbers" for SMS messages. Each SMS we want to send must be assigned a reference byte which is unique for each IMSI that we are handling. In 3gpp world, SMS's are only identified by this reference number, whereas we want to store a unqiue PK for each message. So there's a ton of logic here around allocating and tracking these reference numbers and accounting for retries. Oh, and each message we save in the DB could potentially result in multiple SMS's being sent in the end - which means we need to support assigning multiple reference numbers to each message in the DB.
- Reference numbers are tracked in a separate table where they're FK'd back to the main table for messages. Once we receive a successful delivery report, we clear all the reference numbers for the message. If we get a delivery report that says the message failed to send, we will clear the reference numbers if the message has exceeded the retry limit.

## Test Plan

- New integration test that walks through a workflow with the storage impl. Treat the test case as additional documentation on the detailed semantics of the storage impl throughout different edge cases.

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/cloud/go/services/smsd/storage/sql.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+		}
+		return nil, nil
+	}
+
+	_, err := sqorc.ExecInTx(s.db, nil, nil, txFn)
+	return err
+}
+
+func (s *sqlSMSStorage) ReportDelivery(deliveredMessages map[string][]SMSRef, failedMessages map[string][]SMSFailureReport) error {
+	// TODO: what should we do in this case when we get a delivery digest for
```
Comment (shaddi) [inline lte/cloud/go/services/smsd/storage/sql.go:313]: Semantically these should be ignored.

**Review by shaddi — COMMENTED** [review]
Making the assumption that this is the reasonable way to do storage, this broadly seems fine to me. 

My biggest question is why we need to maintain a second table just for references, since there should always be a 1:1 mapping between SMS and reference number while it is "in flight" (i.e., unacknowledged).

**Comment by xjtian** [comment]
> Making the assumption that this is the reasonable way to do storage, this broadly seems fine to me.
> 
> My biggest question is why we need to maintain a second table just for references, since there should always be a 1:1 mapping between SMS and reference number while it is "in flight" (i.e., unacknowledged).

Each message in the table can decompose into multiple SMS's, so every row in the main table can map to multiple ref nums (per SMS that the message text will be encoded into)

**Review by mpgermano — APPROVED** [review]
Looks like this checks out, though it's a bit hard to digest

**Review by themarwhal — APPROVED** [review]
Only feg changes are go.mod/sum
